app/src/main/java/com/example/antitheft/structure/GPSCommandReceiver.java

```diff
@@ -0,0 +1,49 @@
+package com.example.antitheft.structure;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.eis.smslibrary.SMSMessage;
+import com.eis.smslibrary.listeners.SMSReceivedServiceListener;
+import com.example.antitheft.MainActivity;
+
+/**
+ * This class captures all incoming messages. It extends abstract class
+ * {@link com.eis.smslibrary.listeners.SMSReceivedServiceListener}.
+ *
+ * @author Francesco Bau'
+ * @version 0.1
+ * @since 25/02/2020
+ * @see com.eis.smslibrary.listeners.SMSReceivedServiceListener
+ */
+public class GPSCommandReceiver extends SMSReceivedServiceListener {
+
+    /**
+     * Listener, needed to obtain incoming messages, and to check if they need to be handled as
+     * command or not.
+     *
+     * @param message The incoming message.
+     * @see com.eis.smslibrary.SMSMessage
+     * @see GPSCommandHandler
+     * @see GPSCommandHandler#onCommandReceived(SMSMessage)
+     */
+    @Override
+    public void onMessageReceived(@Nullable SMSMessage message) {
+        if (message == null) return;
+        if (!isValidCommand(message)) return;
+        new GPSCommandHandler().onCommandReceived(message);
```

Comment:
I prefer not to throw an exception, but just to ignore it, because that listener obtaines **all** incoming messages.
Otherwise, if someone sends you a non-valid command via SMS, the app would crash.

Also, I've tried your second proposal, but, even though java is supposed to be "lazy", so if the message is null, then it's supposed to say _"ok, that condition is true, let's skip the other one!"_, , but strangely it doesn't do that for me... I tried that, but the test failed if I tried to give it a null `String` .

So that's why I decided to split the conditions.